nautobot/core/runner/importer.py

```diff
@@ -12,11 +12,6 @@
 from .settings import load_settings, create_module
 
 
-def execfile(afile, globalz=None, localz=None):
```

Comment:
Can you add some context to the PR for the changes made in this file?

Reply:
I did in https://github.com/nautobot/nautobot/pull/5039#discussion_r1439906320, looks like file-level comments don't carry over in the GitHub UI when the file changes. :-/

Reply:
I see now. Do we still get a valid error with an invalid `EXTRA_INSTALLED_APPS` or `EXTRA_MIDDLEWARE` without this code?

Reply:
Invalid would be like the wrong data type (dict instead of list, that sort of thing)? I'd tested earlier with a syntax error in the nautobot_config.py and that seemed fine, but I'm open to doing additional tests.

Reply:
Wrong data type, or a non-existent python library (for `EXTRA_INSTALLED_APPS`). Just want to make sure you captured the `EXTRA_` settings that this custom importer is providing.

Reply:
With `EXTRA_INSTALLED_APPS = ["hello world"]`:

```
nautobot-nautobot-1  | Traceback (most recent call last):
nautobot-nautobot-1  |   File "/usr/local/bin/nautobot-server", line 6, in <module>
nautobot-nautobot-1  |     sys.exit(main())
nautobot-nautobot-1  |   File "/source/nautobot/core/cli/__init__.py", line 53, in main
nautobot-nautobot-1  |     run_app(
nautobot-nautobot-1  |   File "/source/nautobot/core/runner/runner.py", line 297, in run_app
nautobot-nautobot-1  |     management.execute_from_command_line([runner_name, command, *command_args])
nautobot-nautobot-1  |   File "/usr/local/lib/python3.8/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
nautobot-nautobot-1  |     utility.execute()
nautobot-nautobot-1  |   File "/usr/local/lib/python3.8/site-packages/django/core/management/__init__.py", line 395, in execute
nautobot-nautobot-1  |     django.setup()
nautobot-nautobot-1  |   File "/usr/local/lib/python3.8/site-packages/django/__init__.py", line 24, in setup
nautobot-nautobot-1  |     apps.populate(settings.INSTALLED_APPS)
nautobot-nautobot-1  |   File "/usr/local/lib/python3.8/site-packages/django/apps/registry.py", line 91, in populate
nautobot-nautobot-1  |     app_config = AppConfig.create(entry)
nautobot-nautobot-1  |   File "/usr/local/lib/python3.8/site-packages/django/apps/config.py", line 224, in create
nautobot-nautobot-1  |     import_module(entry)
nautobot-nautobot-1  |   File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module
nautobot-nautobot-1  |     return _bootstrap._gcd_import(name[level:], package, level)
nautobot-nautobot-1  |   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
nautobot-nautobot-1  |   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
nautobot-nautobot-1  |   File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked
nautobot-nautobot-1  | ModuleNotFoundError: No module named 'hello world'
```

With `EXTRA_INSTALLED_APPS = {"hello": "world"}`:

```
nautobot-nautobot-1  | Traceback (most recent call last):
nautobot-nautobot-1  |   File "/usr/local/bin/nautobot-server", line 6, in <module>
nautobot-nautobot-1  |     sys.exit(main())
nautobot-nautobot-1  |   File "/source/nautobot/core/cli/__init__.py", line 53, in main
nautobot-nautobot-1  |     run_app(
nautobot-nautobot-1  |   File "/source/nautobot/core/runner/runner.py", line 297, in run_app
nautobot-nautobot-1  |     management.execute_from_command_line([runner_name, command, *command_args])
nautobot-nautobot-1  |   File "/usr/local/lib/python3.8/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
nautobot-nautobot-1  |     utility.execute()
nautobot-nautobot-1  |   File "/usr/local/lib/python3.8/site-packages/django/core/management/__init__.py", line 395, in execute
nautobot-nautobot-1  |     django.setup()
nautobot-nautobot-1  |   File "/usr/local/lib/python3.8/site-packages/django/__init__.py", line 24, in setup
nautobot-nautobot-1  |     apps.populate(settings.INSTALLED_APPS)
nautobot-nautobot-1  |   File "/usr/local/lib/python3.8/site-packages/django/apps/registry.py", line 91, in populate
nautobot-nautobot-1  |     app_config = AppConfig.create(entry)
nautobot-nautobot-1  |   File "/usr/local/lib/python3.8/site-packages/django/apps/config.py", line 224, in create
nautobot-nautobot-1  |     import_module(entry)
nautobot-nautobot-1  |   File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module
nautobot-nautobot-1  |     return _bootstrap._gcd_import(name[level:], package, level)
nautobot-nautobot-1  |   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
nautobot-nautobot-1  |   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
nautobot-nautobot-1  |   File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked
nautobot-nautobot-1  | ModuleNotFoundError: No module named 'hello'
```

Looks like the same stack trace I get in `develop` in both cases -- which makes sense as I think about it, since all the removed code was doing was checking whether the config file could be imported, not whether it's actually semantically valid.

With a syntax error in the config file, `develop` gives:

```
nautobot-nautobot-1  | Traceback (most recent call last):
nautobot-nautobot-1  |   File "/source/nautobot/core/runner/importer.py", line 72, in validate
nautobot-nautobot-1  |     execfile(self.config_path, {"__file__": self.config_path})
nautobot-nautobot-1  |   File "/source/nautobot/core/runner/importer.py", line 17, in execfile
nautobot-nautobot-1  |     exec(fh.read(), globalz, localz)
nautobot-nautobot-1  |   File "<string>", line 74
nautobot-nautobot-1  |     EXTRA_INSTALLED_APPS = {
nautobot-nautobot-1  |                            ^
nautobot-nautobot-1  | SyntaxError: unexpected EOF while parsing
nautobot-nautobot-1  | 
nautobot-nautobot-1  | During handling of the above exception, another exception occurred:
nautobot-nautobot-1  | 
nautobot-nautobot-1  | Traceback (most recent call last):
nautobot-nautobot-1  |   File "/usr/local/bin/nautobot-server", line 6, in <module>
nautobot-nautobot-1  |     sys.exit(main())
nautobot-nautobot-1  |   File "/source/nautobot/core/cli/__init__.py", line 53, in main
nautobot-nautobot-1  |     run_app(
nautobot-nautobot-1  |   File "/source/nautobot/core/runner/runner.py", line 292, in run_app
nautobot-nautobot-1  |     configure_app(config_path=config_path, **kwargs)
nautobot-nautobot-1  |   File "/source/nautobot/core/runner/runner.py", line 131, in configure_app
nautobot-nautobot-1  |     importer.install(
nautobot-nautobot-1  |   File "/source/nautobot/core/runner/importer.py", line 37, in install
nautobot-nautobot-1  |     sys.meta_path.insert(0, LoganImporter(name, config_path, default_settings, **kwargs))
nautobot-nautobot-1  |   File "/source/nautobot/core/runner/importer.py", line 63, in __init__
nautobot-nautobot-1  |     self.validate()
nautobot-nautobot-1  |   File "/source/nautobot/core/runner/importer.py", line 75, in validate
nautobot-nautobot-1  |     raise ConfigurationError(str(e), exc_info[2])
nautobot-nautobot-1  | nautobot.core.runner.importer.ConfigurationError: ('unexpected EOF while parsing (<string>, line 74)', <traceback object at 0x7fd1a1519340>)
```

while this branch instead gives:

```
nautobot-nautobot-1  | Traceback (most recent call last):
nautobot-nautobot-1  |   File "/usr/local/bin/nautobot-server", line 6, in <module>
nautobot-nautobot-1  |     sys.exit(main())
nautobot-nautobot-1  |   File "/source/nautobot/core/cli/__init__.py", line 53, in main
nautobot-nautobot-1  |     run_app(
nautobot-nautobot-1  |   File "/source/nautobot/core/runner/runner.py", line 292, in run_app
nautobot-nautobot-1  |     configure_app(config_path=config_path, **kwargs)
nautobot-nautobot-1  |   File "/source/nautobot/core/runner/runner.py", line 145, in configure_app
nautobot-nautobot-1  |     hasattr(settings, "INSTALLED_APPS")
nautobot-nautobot-1  |   File "/usr/local/lib/python3.8/site-packages/django/conf/__init__.py", line 82, in __getattr__
nautobot-nautobot-1  |     self._setup(name)
nautobot-nautobot-1  |   File "/usr/local/lib/python3.8/site-packages/django/conf/__init__.py", line 69, in _setup
nautobot-nautobot-1  |     self._wrapped = Settings(settings_module)
nautobot-nautobot-1  |   File "/usr/local/lib/python3.8/site-packages/django/conf/__init__.py", line 170, in __init__
nautobot-nautobot-1  |     mod = importlib.import_module(self.SETTINGS_MODULE)
nautobot-nautobot-1  |   File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module
nautobot-nautobot-1  |     return _bootstrap._gcd_import(name[level:], package, level)
nautobot-nautobot-1  |   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
nautobot-nautobot-1  |   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
nautobot-nautobot-1  |   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
nautobot-nautobot-1  |   File "<frozen importlib._bootstrap>", line 655, in _load_unlocked
nautobot-nautobot-1  |   File "<frozen importlib._bootstrap>", line 618, in _load_backward_compatible
nautobot-nautobot-1  |   File "/source/nautobot/core/runner/importer.py", line 115, in load_module
nautobot-nautobot-1  |     load_settings(self.config_path, allow_extras=self.allow_extras, settings=settings_mod)
nautobot-nautobot-1  |   File "/source/nautobot/core/runner/settings.py", line 47, in load_settings
nautobot-nautobot-1  |     spec.loader.exec_module(conf)
nautobot-nautobot-1  |   File "<frozen importlib._bootstrap_external>", line 839, in exec_module
nautobot-nautobot-1  |   File "<frozen importlib._bootstrap_external>", line 976, in get_code
nautobot-nautobot-1  |   File "<frozen importlib._bootstrap_external>", line 906, in source_to_code
nautobot-nautobot-1  |   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
nautobot-nautobot-1  |   File "/opt/nautobot/nautobot_config.py", line 78
nautobot-nautobot-1  |     EXTRA_INSTALLED_APPS = {
nautobot-nautobot-1  |                            ^
nautobot-nautobot-1  | SyntaxError: unexpected EOF while parsing
```

which I actually think is an improvement?

Reply:
Agreed that it's an improvement. I'm happy with it, thanks for testing that out.